plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/index.tsx

```diff
@@ -31,49 +33,62 @@ const Edit = ( props: ProductCollectionEditComponentProps ) => {
 		[ clientId ]
 	);
 
-	const productCollectionUIStateInEditor =
-		getProductCollectionUIStateInEditor( {
-			hasInnerBlocks,
+	const { productCollectionUIStateInEditor, isLoading } =
+		useProductCollectionUIState( {
 			location,
-			attributes: props.attributes,
+			attributes,
+			hasInnerBlocks,
 			usesReference: props.usesReference,
 		} );
 
-	/**
-	 * Component to render based on the UI state.
-	 */
-	let Component,
-		isUsingReferencePreviewMode = false;
-	switch ( productCollectionUIStateInEditor ) {
-		case ProductCollectionUIStatesInEditor.COLLECTION_PICKER:
-			Component = ProductCollectionPlaceholder;
-			break;
-		case ProductCollectionUIStatesInEditor.PRODUCT_REFERENCE_PICKER:
-			Component = ProductPicker;
-			break;
-		case ProductCollectionUIStatesInEditor.VALID:
-			Component = ProductCollectionContent;
-			break;
-		case ProductCollectionUIStatesInEditor.VALID_WITH_PREVIEW:
-			Component = ProductCollectionContent;
-			isUsingReferencePreviewMode = true;
-			break;
-		default:
-			// By default showing collection chooser.
-			Component = ProductCollectionPlaceholder;
+	// Show spinner while calculating Editor UI state.
+	if ( isLoading ) {
+		return (
+			<Flex justify="center" align="center">
+				<Spinner />
+			</Flex>
+		);
 	}
 
+	const productCollectionContentProps: ProductCollectionContentProps = {
+		...props,
+		openCollectionSelectionModal: () => setIsSelectionModalOpen( true ),
+		location,
+		isUsingReferencePreviewMode:
+			productCollectionUIStateInEditor ===
+			ProductCollectionUIStatesInEditor.VALID_WITH_PREVIEW,
+	};
+
+	const renderComponent = () => {
+		switch ( productCollectionUIStateInEditor ) {
+			case ProductCollectionUIStatesInEditor.COLLECTION_PICKER:
+				return <ProductCollectionPlaceholder { ...props } />;
+			case ProductCollectionUIStatesInEditor.PRODUCT_REFERENCE_PICKER:
+			case ProductCollectionUIStatesInEditor.DELETED_PRODUCT_REFERENCE:
+				return (
+					<ProductPicker
+						{ ...props }
+						isDeletedProductReference={
+							productCollectionUIStateInEditor ===
+							ProductCollectionUIStatesInEditor.DELETED_PRODUCT_REFERENCE
+						}
+					/>
+				);
```

Comment:
Super nit: I'm not a fan the situation in which we have two cases pointing to single return but then there's a condition inside taking into account the `case` condition 😅 

So we have:
```
case A:
case B:
    return <ProductPicker prop={ condition === B }
```

while I think we could be more explicit:

```
case A:
   return <ProductPicker prop={ false }
case B:
    return <ProductPicker prop={ true }
```

But it's just some sort of preference hence don't feel it has to be addressed.

Reply:
I don't have a strong opinion so made the changes in 6be9f48 🤝